Description: CVE-2024-36623 - Make progressOutput concurrency safe
 Fix race condition in pkg/streamformatter progressOutput.
 Sync access to the underlying io.Writer with a mutex to prevent
 concurrent write operations resulting in data corruption or crashes.
Origin: upstream, https://github.com/moby/moby/commit/8e3bcf19748838b30e34d612832d1dc9d90363b8
Author: Pawel Gronowski <pawel.gronowski@docker.com>
Last-Update: 2026-01-28

--- a/engine/pkg/streamformatter/streamformatter.go
+++ b/engine/pkg/streamformatter/streamformatter.go
@@ -5,6 +5,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
+	"sync"

 	"github.com/docker/docker/pkg/jsonmessage"
 	"github.com/docker/docker/pkg/progress"
@@ -108,6 +109,7 @@ type progressOutput struct {
 	sf       formatProgress
 	out      io.Writer
 	newLines bool
+	mu       sync.Mutex
 }

 // WriteProgress formats progress information from a ProgressReader.
@@ -119,6 +121,9 @@ func (out *progressOutput) WriteProgress(prog progress.Progress) error {
 		jsonProgress := jsonmessage.JSONProgress{Current: prog.Current, Total: prog.Total, HideCounts: prog.HideCounts, Units: prog.Units}
 		formatted = out.sf.formatProgress(prog.ID, prog.Action, &jsonProgress, prog.Aux)
 	}
+
+	out.mu.Lock()
+	defer out.mu.Unlock()
 	_, err := out.out.Write(formatted)
 	if err != nil {
 		return err
Description: CVE-2024-29018 - Do not forward DNS requests to external servers for internal networks
 Moby's networking implementation allows creation of internal networks using
 the --internal flag. Containers on internal networks are supposed to be
 isolated from external communications. However, the embedded DNS server was
 forwarding DNS requests to external nameservers even for containers on
 internal networks. This could allow data exfiltration via DNS queries.
 .
 This patch adds a SetForwardingPolicy method to the Resolver interface and
 implementation. When a container joins or leaves a network, the forwarding
 policy is updated based on whether the sandbox has external connectivity
 (gateway endpoint). If no gateway exists (internal network), DNS forwarding
 is disabled.
 .
 This is a backport for v20.10 adapted from the upstream fix for v26+.
Origin: upstream, https://github.com/moby/moby/commit/e63daec8672d77ac0b2b5c262ef525c7cf17fd20
Bug: https://github.com/moby/moby/security/advisories/GHSA-mq39-4gv4-mvpx
Author: Pawel Gronowski <pawel.gronowski@docker.com>
Forwarded: not-needed
Last-Update: 2026-01-28

--- a/libnetwork/resolver.go
+++ b/libnetwork/resolver.go
@@ -32,6 +32,9 @@ type Resolver interface {
 	SetExtServers([]extDNSEntry)
 	// ResolverOptions returns resolv.conf options that should be set
 	ResolverOptions() []string
+	// SetForwardingPolicy re-configures the embedded DNS resolver to either
+	// enable or disable forwarding DNS queries to external servers.
+	SetForwardingPolicy(policy bool)
 }

 // DNSBackend represents a backend DNS resolver used for DNS name
@@ -196,6 +199,12 @@ func (r *resolver) SetExtServers(extDNS []extDNSEntry) {
 	}
 }

+// SetForwardingPolicy re-configures the embedded DNS resolver to either enable
+// or disable forwarding DNS queries to external servers.
+func (r *resolver) SetForwardingPolicy(policy bool) {
+	r.proxyDNS = policy
+}
+
 func (r *resolver) NameServer() string {
 	return r.listenAddress
 }
--- a/libnetwork/endpoint.go
+++ b/libnetwork/endpoint.go
@@ -553,7 +553,13 @@ func (ep *endpoint) sbJoin(sb *sandbox, options ...EndpointOption) (err error) {
 		return sb.setupDefaultGW()
 	}

-	moveExtConn := sb.getGatewayEndpoint() != extEp
+	currentExtEp := sb.getGatewayEndpoint()
+	// Enable upstream forwarding if the sandbox gained external connectivity.
+	if sb.resolver != nil {
+		sb.resolver.SetForwardingPolicy(currentExtEp != nil)
+	}
+
+	moveExtConn := currentExtEp != extEp

 	if moveExtConn {
 		if extEp != nil {
@@ -789,6 +795,11 @@ func (ep *endpoint) sbLeave(sb *sandbox, force bool, options ...EndpointOption)

 	// New endpoint providing external connectivity for the sandbox
 	extEp = sb.getGatewayEndpoint()
+	// Disable upstream forwarding if the sandbox lost external connectivity.
+	if sb.resolver != nil {
+		sb.resolver.SetForwardingPolicy(extEp != nil)
+	}
+
 	if moveExtConn && extEp != nil {
 		logrus.Debugf("Programming external connectivity on endpoint %s (%s)", extEp.Name(), extEp.ID())
 		extN, err := extEp.getNetworkFromStore()
